File: frontend/src/types/review.ts
/**
 * Review workflow types for document collaboration.
 */

export type ReviewType = 'feedback' | 'approval' | 'peer_review' | 'editorial';
export type ReviewStatus = 'pending' | 'in_progress' | 'changes_requested' | 'approved' | 'completed' | 'cancelled';
export type ReviewPriority = 'low' | 'normal' | 'high' | 'urgent';
export type ReviewDecision = 'approved' | 'rejected' | 'needs_revision';
export type AssignmentStatus = 'pending' | 'accepted' | 'declined' | 'in_progress' | 'completed';
export type AssignmentRole = 'reviewer' | 'primary_reviewer' | 'approver';
export type ReviewerRecommendation = 'approve' | 'reject' | 'revise' | 'abstain';
export type CommentType = 'general' | 'inline' | 'suggestion' | 'question' | 'issue' | 'gap_identified' | 'clarity_needed' | 'methodology_concern' | 'consistency_issue';
export type CommentSeverity = 'critical' | 'major' | 'minor' | 'suggestion';
export type CommentSource = 'human' | 'ai_suggestion' | 'ai_accepted' | 'ai_dismissed';

export interface ReviewAssignment {
  id: string;
  review_id: string;
  reviewer_id: string;
  assigned_by_id: string | null;
  status: AssignmentStatus;
  role: AssignmentRole;
  responded_at: string | null;
  completed_at: string | null;
  recommendation: ReviewerRecommendation | null;
  notes: string | null;
  due_date: string | null;
  created_at: string;
  updated_at: string;
}

export interface ReviewComment {
  id: string;
  review_id: string;
  user_id: string;
  content: string;
  comment_type: CommentType;
  selected_text: string | null;
  anchor_data: ReviewCommentAnchor | null;
  severity: CommentSeverity | null;
  is_resolved: boolean;
  resolved_by_id: string | null;
  resolved_at: string | null;
  resolution_notes: string | null;
  parent_comment_id: string | null;
  edited_at: string | null;
  created_at: string;
  updated_at: string;
  replies: ReviewComment[];
  // AI-specific fields
  source: CommentSource;
  ai_confidence: number | null;
  question_for_author: string | null;
  why_this_matters: string | null;
}

/** Anchor data for positioning comments in documents/tasks */
export interface ReviewCommentAnchor {
  source_type: 'task' | 'document';
  source_id: string;
  document_title?: string;
  paragraph?: number;
  text_snippet?: string;
  offset?: number;
  [key: string]: unknown;
}

export interface Review {
  id: string;
  document_id: string;
  project_id: string;
  title: string;
  description: string | null;
  review_type: ReviewType;
  status: ReviewStatus;
  priority: ReviewPriority;
  document_version: number;
  requested_by_id: string | null;
  due_date: string | null;
  completed_at: string | null;
  completed_by_id: string | null;
  decision: ReviewDecision | null;
  decision_notes: string | null;
  tags: string[];
  created_at: string;
  updated_at: string;
  assignments: ReviewAssignment[];
}

export interface ReviewListResponse {
  items: Review[];
  total: number;
  page: number;
  page_size: number;
}

export interface ReviewStats {
  total_comments: number;
  resolved_comments: number;
  unresolved_comments: number;
  total_reviewers: number;
  completed_reviews: number;
  pending_reviews: number;
  completion_percentage: number;
}

// Request types
export interface CreateReviewRequest {
  document_id: string;
  project_id: string;
  title: string;
  description?: string;
  review_type?: ReviewType;
  priority?: ReviewPriority;
  due_date?: string;
  reviewer_ids?: string[];
  tags?: string[];
}

export interface UpdateReviewRequest {
  title?: string;
  description?: string;
  status?: ReviewStatus;
  priority?: ReviewPriority;
  due_date?: string;
  decision?: ReviewDecision;
  decision_notes?: string;
  tags?: string[];
}

export interface CreateAssignmentRequest {
  reviewer_id: string;
  role?: AssignmentRole;
  due_date?: string;
}

export interface UpdateAssignmentRequest {
  status?: AssignmentStatus;
  recommendation?: ReviewerRecommendation;
  notes?: string;
}

export interface CreateCommentRequest {
  content: string;
  comment_type?: CommentType;
  selected_text?: string;
  anchor_data?: ReviewCommentAnchor;
  severity?: CommentSeverity;
  parent_comment_id?: string;
  // AI-specific fields (for programmatic creation)
  source?: CommentSource;
  ai_confidence?: number;
  question_for_author?: string;
  why_this_matters?: string;
}

export interface UpdateCommentRequest {
  content?: string;
  is_resolved?: boolean;
  resolution_notes?: string;
}

/** Request to update AI suggestion status */
export interface UpdateAISuggestionRequest {
  action: 'accept' | 'dismiss';
  resolution_notes?: string;
}

/** Request to bulk update AI suggestions */
export interface BulkUpdateAISuggestionsRequest {
  comment_ids: string[];
  action: 'accept' | 'dismiss';
}

// Filter types
export interface ReviewFilters {
  project_id?: string;
  document_id?: string;
  status?: ReviewStatus;
  review_type?: ReviewType;
  assigned_to_me?: boolean;
  requested_by_me?: boolean;
  page?: number;
  page_size?: number;
}

// Helper constants
export const REVIEW_STATUS_LABELS: Record<ReviewStatus, string> = {
  pending: 'Pending',
  in_progress: 'In Progress',
  changes_requested: 'Changes Requested',
  approved: 'Approved',
  completed: 'Completed',
  cancelled: 'Cancelled',
};

export const REVIEW_STATUS_COLORS: Record<ReviewStatus, string> = {
  pending: 'bg-yellow-100 text-yellow-800',
  in_progress: 'bg-blue-100 text-blue-800',
  changes_requested: 'bg-orange-100 text-orange-800',
  approved: 'bg-green-100 text-green-800',
  completed: 'bg-gray-100 text-gray-800',
  cancelled: 'bg-red-100 text-red-800',
};

export const REVIEW_PRIORITY_LABELS: Record<ReviewPriority, string> = {
  low: 'Low',
  normal: 'Normal',
  high: 'High',
  urgent: 'Urgent',
};

export const REVIEW_PRIORITY_COLORS: Record<ReviewPriority, string> = {
  low: 'bg-gray-100 text-gray-800',
  normal: 'bg-blue-100 text-blue-800',
  high: 'bg-orange-100 text-orange-800',
  urgent: 'bg-red-100 text-red-800',
};

export const REVIEW_TYPE_LABELS: Record<ReviewType, string> = {
  feedback: 'Feedback',
  approval: 'Approval',
  peer_review: 'Peer Review',
  editorial: 'Editorial',
};

export const COMMENT_TYPE_LABELS: Record<CommentType, string> = {
  general: 'General',
  inline: 'Inline',
  suggestion: 'Suggestion',
  question: 'Question',
  issue: 'Issue',
  gap_identified: 'Gap Identified',
  clarity_needed: 'Clarity Needed',
  methodology_concern: 'Methodology Concern',
  consistency_issue: 'Consistency Issue',
};

export const COMMENT_TYPE_ICONS: Record<CommentType, string> = {
  general: '💬',
  inline: '📍',
  suggestion: '💡',
  question: '❓',
  issue: '⚠️',
  gap_identified: '🔍',
  clarity_needed: '🔮',
  methodology_concern: '⚙️',
  consistency_issue: '🔄',
};

export const COMMENT_SEVERITY_LABELS: Record<CommentSeverity, string> = {
  critical: 'Critical',
  major: 'Major',
  minor: 'Minor',
  suggestion: 'Suggestion',
};

export const COMMENT_SEVERITY_COLORS: Record<CommentSeverity, string> = {
  critical: 'bg-red-100 text-red-800',
  major: 'bg-orange-100 text-orange-800',
  minor: 'bg-yellow-100 text-yellow-800',
  suggestion: 'bg-blue-100 text-blue-800',
};

// ============================================================================
// Workflow Integration Types (Task-Review Integration)
// ============================================================================

export type WorkflowStage = 'not_started' | 'in_progress' | 'under_review' | 'review_approved' | 'review_rejected' | 'completed' | 'unknown';
export type OverallReviewStatus = 'none' | 'pending' | 'approved' | 'rejected' | 'mixed';

/** Review summary within task review status */
export interface ReviewSummary {
  id: string;
  document_id: string;
  title: string;
  status: ReviewStatus;
  decision: ReviewDecision | null;
}

/** Aggregate review status for a task */
export interface TaskReviewStatus {
  total_reviews: number;
  pending_reviews: number;
  approved_reviews: number;
  rejected_reviews: number;
  all_approved: boolean;
  overall_status: OverallReviewStatus;
  reviews: ReviewSummary[];
}

/** Complete workflow state for a task */
export interface TaskWorkflowState {
  task_id: string;
  task_status: string;
  task_title: string;
  linked_documents: number;
  reviewable_documents: number;
  assignees: number;
  review_status: TaskReviewStatus;
  can_submit_for_review: boolean;
  submit_blocked_reason: string | null;
  workflow_stage: WorkflowStage;
}

/** A unified work item (task or review assignment) */
export interface WorkItem {
  type: 'task' | 'review';
  id: string;
  title: string;
  status: string;
  priority: string;
  due_date: string | null;
  assignment_status: string;
  assignment_role: string;
  project_id: string;
  created_at: string;
  // Review-specific fields
  document_id?: string;
  task_id?: string;
}

/** Response from /tasks/my/work-items endpoint */
export interface WorkItemsResponse {
  tasks: WorkItem[];
  reviews: WorkItem[];
  combined: WorkItem[];
  total_tasks: number;
  total_reviews: number;
}

/** Request to submit a task for review */
export interface SubmitForReviewRequest {
  reviewer_ids?: string[];
  review_type?: ReviewType;
  priority?: ReviewPriority;
  due_date?: string;
  auto_transition_task?: boolean;
}

/** Response from submit-for-review endpoint */
export interface SubmitForReviewResponse {
  id: string;
  document_id: string;
  title: string;
  status: ReviewStatus;
}

// Workflow status labels
export const WORKFLOW_STAGE_LABELS: Record<WorkflowStage, string> = {
  not_started: 'Not Started',
  in_progress: 'In Progress',
  under_review: 'Under Review',
  review_approved: 'Review Approved',
  review_rejected: 'Review Rejected',
  completed: 'Completed',
  unknown: 'Unknown',
};

export const WORKFLOW_STAGE_COLORS: Record<WorkflowStage, string> = {
  not_started: 'bg-gray-100 text-gray-800 dark:bg-gray-700 dark:text-gray-200',
  in_progress: 'bg-blue-100 text-blue-800 dark:bg-blue-900/30 dark:text-blue-300',
  under_review: 'bg-yellow-100 text-yellow-800 dark:bg-yellow-900/30 dark:text-yellow-300',
  review_approved: 'bg-green-100 text-green-800 dark:bg-green-900/30 dark:text-green-300',
  review_rejected: 'bg-red-100 text-red-800 dark:bg-red-900/30 dark:text-red-300',
  completed: 'bg-emerald-100 text-emerald-800 dark:bg-emerald-900/30 dark:text-emerald-300',
  unknown: 'bg-gray-100 text-gray-500 dark:bg-gray-700 dark:text-gray-400',
};

export const OVERALL_REVIEW_STATUS_LABELS: Record<OverallReviewStatus, string> = {
  none: 'No Reviews',
  pending: 'Pending',
  approved: 'Approved',
  rejected: 'Rejected',
  mixed: 'Mixed',
};

export const OVERALL_REVIEW_STATUS_COLORS: Record<OverallReviewStatus, string> = {
  none: 'bg-gray-100 text-gray-500 dark:bg-gray-700 dark:text-gray-400',
  pending: 'bg-yellow-100 text-yellow-800 dark:bg-yellow-900/30 dark:text-yellow-300',
  approved: 'bg-green-100 text-green-800 dark:bg-green-900/30 dark:text-green-300',
  rejected: 'bg-red-100 text-red-800 dark:bg-red-900/30 dark:text-red-300',
  mixed: 'bg-orange-100 text-orange-800 dark:bg-orange-900/30 dark:text-orange-300',
};

// ============================================================================
// AI Review Types and Labels
// ============================================================================

export const COMMENT_SOURCE_LABELS: Record<CommentSource, string> = {
  human: 'Human',
  ai_suggestion: 'AI Suggestion',
  ai_accepted: 'AI Accepted',
  ai_dismissed: 'AI Dismissed',
};

export const COMMENT_SOURCE_COLORS: Record<CommentSource, string> = {
  human: 'bg-gray-100 text-gray-800 dark:bg-gray-700 dark:text-gray-200',
  ai_suggestion: 'bg-purple-100 text-purple-800 dark:bg-purple-900/30 dark:text-purple-300',
  ai_accepted: 'bg-green-100 text-green-800 dark:bg-green-900/30 dark:text-green-300',
  ai_dismissed: 'bg-gray-100 text-gray-500 dark:bg-gray-700 dark:text-gray-400',
};

/** AI comment types that represent substantive review issues */
export const AI_COMMENT_TYPES: CommentType[] = [
  'gap_identified',
  'clarity_needed',
  'methodology_concern',
  'consistency_issue',
];

/** Check if a comment is an AI-generated suggestion */
export const isAISuggestion = (comment: ReviewComment): boolean => {
  return comment.source === 'ai_suggestion';
};

/** Check if a comment has AI fields populated */
export const hasAIContent = (comment: ReviewComment): boolean => {
  return !!(comment.question_for_author || comment.why_this_matters);
};

/** Get the display badge for a comment source */
export const getCommentSourceBadge = (source: CommentSource): { label: string; color: string; icon: string } => {
  const badges: Record<CommentSource, { label: string; color: string; icon: string }> = {
    human: { label: 'Human', color: 'bg-gray-100 text-gray-700', icon: '👤' },
    ai_suggestion: { label: 'AI', color: 'bg-purple-100 text-purple-700', icon: '🤖' },
    ai_accepted: { label: 'Addressed', color: 'bg-green-100 text-green-700', icon: '✅' },
    ai_dismissed: { label: 'Dismissed', color: 'bg-gray-100 text-gray-500', icon: '➖' },
  };
  return badges[source];
};

File: backend/researchhub/models/review.py
"""Review workflow models for document collaboration."""

from datetime import datetime
from typing import TYPE_CHECKING
from uuid import UUID

from sqlalchemy import Boolean, DateTime, Float, ForeignKey, Integer, String, Text
from sqlalchemy.dialects.postgresql import ARRAY, JSONB, UUID as PGUUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from researchhub.db.base import BaseModel

if TYPE_CHECKING:
    from researchhub.models.document import Document
    from researchhub.models.project import Project, Task
    from researchhub.models.user import User


class Review(BaseModel):
    """Review request for a document."""

    __tablename__ = "reviews"

    # Document being reviewed
    document_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("documents.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    # Project context
    project_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("projects.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    # Optional link to originating task (for task-review integration)
    task_id: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("tasks.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )

    # Should review completion auto-update task status?
    auto_transition_task: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=True
    )

    # Review title and description
    title: Mapped[str] = mapped_column(String(500), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Review type
    review_type: Mapped[str] = mapped_column(
        String(50), nullable=False, default="feedback"
    )  # feedback, approval, peer_review, editorial

    # Review status
    status: Mapped[str] = mapped_column(
        String(50), nullable=False, default="pending"
    )  # pending, in_progress, changes_requested, approved, completed, cancelled

    # Priority
    priority: Mapped[str] = mapped_column(
        String(50), nullable=False, default="normal"
    )  # low, normal, high, urgent

    # Document version being reviewed (snapshot)
    document_version: Mapped[int] = mapped_column(Integer, nullable=False, default=1)

    # Requester
    requested_by_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )

    # Due date
    due_date: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )

    # Completion
    completed_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    completed_by_id: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )

    # Final decision
    decision: Mapped[str | None] = mapped_column(
        String(50), nullable=True
    )  # approved, rejected, needs_revision

    # Decision notes
    decision_notes: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Tags
    tags: Mapped[list[str]] = mapped_column(
        ARRAY(String), nullable=False, server_default="{}"
    )

    # Settings
    settings: Mapped[dict] = mapped_column(JSONB, nullable=False, default=dict)

    # Relationships
    document: Mapped["Document"] = relationship("Document")
    project: Mapped["Project"] = relationship("Project")
    task: Mapped["Task | None"] = relationship("Task")
    requested_by: Mapped["User | None"] = relationship(
        "User", foreign_keys=[requested_by_id]
    )
    completed_by: Mapped["User | None"] = relationship(
        "User", foreign_keys=[completed_by_id]
    )
    assignments: Mapped[list["ReviewAssignment"]] = relationship(
        "ReviewAssignment", back_populates="review", lazy="selectin"
    )
    comments: Mapped[list["ReviewComment"]] = relationship(
        "ReviewComment", back_populates="review", lazy="selectin"
    )

    def __repr__(self) -> str:
        try:
            return f"<Review {self.title[:30]} on doc={self.document_id}>"
        except Exception:
            return f"<Review id={self.id}>"


class ReviewAssignment(BaseModel):
    """Assignment of a reviewer to a review."""

    __tablename__ = "review_assignments"

    review_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("reviews.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    # Assigned reviewer
    reviewer_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    # Who made the assignment
    assigned_by_id: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )

    # Assignment status
    status: Mapped[str] = mapped_column(
        String(50), nullable=False, default="pending"
    )  # pending, accepted, declined, in_progress, completed

    # Reviewer's role
    role: Mapped[str] = mapped_column(
        String(50), nullable=False, default="reviewer"
    )  # reviewer, primary_reviewer, approver

    # Response tracking
    responded_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )

    # Completion
    completed_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )

    # Reviewer's recommendation
    recommendation: Mapped[str | None] = mapped_column(
        String(50), nullable=True
    )  # approve, reject, revise, abstain

    # Reviewer's notes
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Optional deadline for this reviewer
    due_date: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )

    # Relationships
    review: Mapped["Review"] = relationship("Review", back_populates="assignments")
    reviewer: Mapped["User"] = relationship("User", foreign_keys=[reviewer_id])
    assigned_by: Mapped["User | None"] = relationship("User", foreign_keys=[assigned_by_id])

    def __repr__(self) -> str:
        return f"<ReviewAssignment review={self.review_id} reviewer={self.reviewer_id}>"


class ReviewComment(BaseModel):
    """Comment on a review (can be anchored to document text)."""

    __tablename__ = "review_comments"

    review_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("reviews.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    # Comment author
    user_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
    )

    # Comment content
    content: Mapped[str] = mapped_column(Text, nullable=False)

    # Comment type
    comment_type: Mapped[str] = mapped_column(
        String(50), nullable=False, default="general"
    )  # general, inline, suggestion, question, issue, gap_identified, clarity_needed, methodology_concern, consistency_issue

    # For inline comments - anchor to document text
    selected_text: Mapped[str | None] = mapped_column(Text, nullable=True)
    anchor_data: Mapped[dict | None] = mapped_column(JSONB, nullable=True)

    # Source tracking for AI vs human comments
    source: Mapped[str] = mapped_column(
        String(50), nullable=False, default="human"
    )  # human, ai_suggestion, ai_accepted, ai_dismissed

    # AI-specific fields
    ai_confidence: Mapped[float | None] = mapped_column(Float, nullable=True)
    question_for_author: Mapped[str | None] = mapped_column(Text, nullable=True)
    why_this_matters: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Severity/category for issues
    severity: Mapped[str | None] = mapped_column(
        String(50), nullable=True
    )  # critical, major, minor, suggestion

    # Resolution status
    is_resolved: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    resolved_by_id: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )
    resolved_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    resolution_notes: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Threading
    parent_comment_id: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("review_comments.id", ondelete="CASCADE"),
        nullable=True,
    )

    # Edit tracking
    edited_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )

    # Relationships
    review: Mapped["Review"] = relationship("Review", back_populates="comments")
    user: Mapped["User"] = relationship("User", foreign_keys=[user_id])
    resolved_by: Mapped["User | None"] = relationship("User", foreign_keys=[resolved_by_id])
    parent_comment: Mapped["ReviewComment | None"] = relationship(
        "ReviewComment", remote_side="ReviewComment.id", back_populates="replies"
    )
    replies: Mapped[list["ReviewComment"]] = relationship(
        "ReviewComment", back_populates="parent_comment", lazy="selectin"
    )

    def __repr__(self) -> str:
        return f"<ReviewComment {self.id} on review={self.review_id}>"


class AutoReviewConfig(BaseModel):
    """Per-organization configuration for AI auto-reviews."""

    __tablename__ = "auto_review_configs"

    # Organization this config belongs to
    organization_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("organizations.id", ondelete="CASCADE"),
        nullable=False,
        unique=True,
        index=True,
    )

    # Trigger settings
    on_document_create: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False
    )
    on_document_update: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False
    )
    on_task_submit_review: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=True
    )

    # Focus areas for AI review (e.g., ["methodology", "clarity", "completeness"])
    default_focus_areas: Mapped[list[str]] = mapped_column(
        ARRAY(String), nullable=False, server_default="{}"
    )

    # Minimum document length (in characters) to trigger auto-review
    min_document_length: Mapped[int] = mapped_column(
        Integer, nullable=False, default=100
    )

    # Cooldown period in hours before re-reviewing same content
    review_cooldown_hours: Mapped[int] = mapped_column(
        Integer, nullable=False, default=24
    )

    # Maximum number of AI suggestions per review
    max_suggestions_per_review: Mapped[int] = mapped_column(
        Integer, nullable=False, default=10
    )

    # Whether AI should auto-create review when triggered
    auto_create_review: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=True
    )

    # Who created/updated this config
    updated_by_id: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )

    # Relationships
    updated_by: Mapped["User | None"] = relationship("User")

    def __repr__(self) -> str:
        return f"<AutoReviewConfig org={self.organization_id}>"


class AutoReviewLog(BaseModel):
    """Log of auto-reviews performed to prevent duplicate processing."""

    __tablename__ = "auto_review_logs"

    # What was reviewed
    task_id: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("tasks.id", ondelete="CASCADE"),
        nullable=True,
        index=True,
    )
    document_id: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("documents.id", ondelete="CASCADE"),
        nullable=True,
        index=True,
    )

    # Content hash to detect duplicate content
    content_hash: Mapped[str] = mapped_column(String(64), nullable=False, index=True)

    # Review that was created
    review_id: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("reviews.id", ondelete="SET NULL"),
        nullable=True,
    )

    # Number of suggestions generated
    suggestions_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)

    # Trigger source
    trigger_source: Mapped[str] = mapped_column(
        String(50), nullable=False
    )  # document_create, document_update, task_submit_review, manual

    # Processing status
    status: Mapped[str] = mapped_column(
        String(50), nullable=False, default="pending"
    )  # pending, processing, completed, failed

    # Error message if failed
    error_message: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Processing timestamps
    started_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    completed_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )

    def __repr__(self) -> str:
        return f"<AutoReviewLog task={self.task_id} doc={self.document_id}>"
